### PR TITLE
OPENEUROPA-3373: Add query functionality

### PR DESCRIPTION
**Pull request by hernani**
### Description

[Insert description here]

### Change log

- Added:
- Changed:
- Deprecated:
- Removed:
- Fixed:
- Security:

### Commands

```sh
[Insert commands here]

```

### CODE REVIEW

File: src/EventSubscriber/QuerySubscriber.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+   *
+   * @param \Drupal\search_api\Event\QueryPreExecuteEvent $event
+   *   The query alter event.
+   */
+  public function queryAlter(QueryPreExecuteEvent $event) {
+    $query = $event->getQuery();
+
+    $ignored_filters = $preset_filters = [];
+
+    if ($query->getIndex()->getServerInstance()->supportsFeature('search_api_facets')) {
```
Comment (upchuk) [inline src/EventSubscriber/QuerySubscriber.php:34]: Can we please return early if it doesn't support the feature? makes it a bit more readable.

File: src/EventSubscriber/QuerySubscriber.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      /** @var \Drupal\facets\FacetManager\DefaultFacetManager $facet_manager */
+      $facet_manager = \Drupal::service('facets.manager');
+      $queryTypePluginManager = \Drupal::service('plugin.manager.facets.query_type');
```
Comment (upchuk) [inline src/EventSubscriber/QuerySubscriber.php:37]: snake_case

File: src/EventSubscriber/QuerySubscriber.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+use Drupal\search_api\Event\QueryPreExecuteEvent;
+use Drupal\search_api\Event\SearchApiEvents;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Provides an event subscriber that allows to alter list source queries.
+ */
+class QuerySubscriber implements EventSubscriberInterface {
+
+  /**
```
Comment (upchuk) [inline src/EventSubscriber/QuerySubscriber.php:14]: Please inject all dependencies.

File: src/Plugin/facets/facet_source/ListFacetSource.php [hunk 4]
```diff
@@ -58,7 +59,8 @@ public function __construct(array $configuration, $plugin_id, array $plugin_defi
     $this->currentPathStack = $current_path_stack;
     $this->requestStack = $request_stack;
     $this->entityTypeManager = $entity_type_manager;
-    $this->index = $plugin_definition['index'];
+    // This needs to be loaded as search api doesn't rely on getIndex().
+    $this->index = $this->entityTypeManager->getStorage('search_api_index')->load($plugin_definition['index']);
```
Comment (upchuk) [inline src/Plugin/facets/facet_source/ListFacetSource.php:63]: Why are you doing this? Not all the times this plugin needs to be instantiated will you need the index loaded. So load the index only when you need it. It was correct before.

File: src/Plugin/search_api/display/ListDisplayDeriver.php [hunk 5]
```diff
@@ -26,8 +26,8 @@ public function getDerivativeDefinitions($base_plugin_definition) {
       foreach ($datasources as $datasource) {
         $entity_type = $datasource->getEntityTypeId();
         $bundles = $datasource->getBundles();
-        foreach ($bundles as $bundle_id => $label) {
-          $id = $entity_type . PluginBase::DERIVATIVE_SEPARATOR . $bundle_id;
+        foreach ($bundles as $id => $label) {
+          $id = $entity_type . PluginBase::DERIVATIVE_SEPARATOR . $id;
```
Comment (upchuk) [inline src/Plugin/search_api/display/ListDisplayDeriver.php:30]: What is this change? Leave `bundle_id`. Yo uare using the variable $id in the foreach and then overriding it inside the loop...

File: src/ListSource.php [hunk 6]
```diff
@@ -106,4 +107,23 @@ public function getIndex(): IndexInterface {
...
+
+    $query = $this->index->query([
+      'limit' => $limit,
+      'offset' => ($limit * $page),
+    ]);
+
+    $query_options = new ListQueryOptions($ignored_filters, $preset_filters_values);
+    $query->setOption('oe_list_page_query_options', $query_options);
+    $query->setSearchId($this->getSearchId());
+    $query->addCondition('type', $this->getBundle());
```
Comment (hernani) [inline src/ListSource.php:123]: To be fixed.

File: src/Plugin/facets/facet_source/ListFacetSource.php (hunk 4 above)
Comment (hernani) [reply src/Plugin/facets/facet_source/ListFacetSource.php:63]: Search API does not always use the getIndex method to access this property..

**Pull request merged**